compiler/optimizer/demand.go

```diff
@@ -0,0 +1,177 @@
+package optimizer
+
+import (
+	//"fmt"
+	"github.com/brimdata/zed/compiler/ast/dag"
+)
+
+type Demand interface {
+	isDemand()
+}
+
+// TODO Think about normal form for Demand.
+
+// A nil Demand means 'demand nothing'.
+type DemandAll struct{}
+type DemandKey struct {
+	Key   string
+	Value Demand // Not nil
+}
+type DemandUnion [2]Demand // Disjoint. Not nil.
+
+func (demand DemandAll) isDemand()   {}
+func (demand DemandKey) isDemand()   {}
+func (demand DemandUnion) isDemand() {}
+
+func demandKey(key string, value Demand) Demand {
+	if value == nil {
+		return nil
+	}
+	return DemandKey{
+		Key:   key,
+		Value: value,
+	}
+}
+
+func demandUnion(a Demand, b Demand) Demand {
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+
+	if _, ok := a.(DemandAll); ok {
+		return a
+	}
+	if _, ok := b.(DemandAll); ok {
+		return b
+	}
+
+	if a, ok := a.(DemandKey); ok {
+		if b, ok := b.(DemandKey); ok {
+			if a.Key == b.Key {
+				return DemandKey{
+					Key:   a.Key,
+					Value: demandUnion(a.Value, b.Value),
+				}
+			}
+		}
+	}
+
+	return DemandUnion([2]Demand{a, b})
+}
+
+func demandForSeq(seq dag.Seq) map[*dag.Op]Demand {
+	demands := make(map[*dag.Op]Demand)
+	demandForSeqInto(demands, DemandAll{}, seq)
+
+	//walk(seq, true, func(seq dag.Seq) dag.Seq {
+	//    for i := range seq {
+	//        fmt.Println(seq[i], " ", demands[&seq[i]])
+	//    }
+	//    return seq
+	//})
+
+	return demands
+}
+
+func demandForSeqInto(demands map[*dag.Op]Demand, demandOnSeq Demand, seq dag.Seq) {
+	var demand = demandOnSeq
+	for i := len(seq) - 1; i >= 0; i-- {
+		op_ptr := &seq[i]
```

Comment:
Don't you want the dag.Op interface value (seq[i]) rather than the address of the slot in the slice?